From 6b30a48f860dc3e040fd7dc937bddd003b3ae8ba Mon Sep 17 00:00:00 2001
From: Jason Macnak <natsu@google.com>
Date: Thu, 20 Feb 2025 09:25:19 -0800
Subject: [PATCH] Fix LLVM's AlignOf after previous change to support c++23

https://swiftshader-review.googlesource.com/c/SwiftShader/+/74809
seems to have dropped the `T` arg from the alignof and sizeof
calculations which resulted in build errors in AOSP. See the
checks tab on aosp/3503612.

Bug: b/397950401
Test: build in Android repo
Change-Id: I555e2b59181a6d63603191a90b4d6c1bb2080efd
Reviewed-on: https://swiftshader-review.googlesource.com/c/SwiftShader/+/75068
Reviewed-by: Shahbaz Youssefi <syoussefi@google.com>
Commit-Queue: Shahbaz Youssefi <syoussefi@google.com>
Tested-by: Shahbaz Youssefi <syoussefi@google.com>
Kokoro-Result: kokoro <noreply+kokoro@google.com>
---
 third_party/llvm-16.0/llvm/include/llvm/Support/AlignOf.h | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/third_party/llvm-16.0/llvm/include/llvm/Support/AlignOf.h b/third_party/llvm-16.0/llvm/include/llvm/Support/AlignOf.h
index a8c45785c..da1d7e3b5 100644
--- a/third_party/llvm-16.0/llvm/include/llvm/Support/AlignOf.h
+++ b/third_party/llvm-16.0/llvm/include/llvm/Support/AlignOf.h
@@ -20,7 +20,7 @@ namespace llvm {
 /// A suitably aligned and sized character array member which can hold elements
 /// of any type.
 template <typename T, typename... Ts> struct AlignedCharArrayUnion {
-  alignas(Ts...) char buffer[std::max({sizeof(Ts)...})];
+  alignas(std::max({std::size_t(1), alignof(T), alignof(Ts)...})) char buffer[std::max({std::size_t(1), sizeof(T), sizeof(Ts)...})];
 };
 
 } // end namespace llvm
-- 
2.39.2

